src/Psl/Type/Internal/MutableSetType.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use Psl\Collection;
+use Psl\Str;
+use Psl\Type;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Throwable;
+
+use function is_iterable;
+use function is_object;
+
+/**
+ * @template T of array-key
+ *
+ * @extends Type\Type<Collection\MutableSetInterface<T>>
+ *
+ * @internal
+ */
+final readonly class MutableSetType extends Type\Type
+{
+    /**
+     * @psalm-mutation-free
+     *
+     * @param Type\TypeInterface<T> $key_type
+     */
+    public function __construct(
+        private readonly Type\TypeInterface $key_type,
+    ) {
+    }
+
+    /**
+     * @throws CoercionException
+     *
+     * @return Collection\MutableSetInterface<T>
+     */
+    public function coerce(mixed $value): Collection\MutableSetInterface
+    {
+        if (is_iterable($value)) {
+            /** @var Type\Type<T> $key_type */
+            $key_type = $this->key_type;
+            /** @var list<T> $keys */
+            $keys = [];
+            $key = null;
+            $iterating = true;
+            try {
+                /**
+                 * @var T $key
+                 */
+                foreach ($value as $key => $_) {
```

Comment:
Using the key doesn't feel very natural here.
Wouldnt it make more sense to use the values instead? If one wanted to use the keys, they could first grab them and pass those along to the type? (same for the other type)

Reply:
updated.